torchtune/datasets/_packed.py

```diff
@@ -119,156 +117,192 @@ def _pack(self) -> None:
+
             # If the dataset outputs samples that are larger than the specified
             # max_seq_len and we're unable to split it, user needs to modify
             # one of the two parameters
             seq_len = len(tokens)
             if seq_len > self.max_seq_len and not self.split_across_pack:
                 raise ValueError(
-                    f"Dataset sample is too long ({len(tokens)} > {self.max_seq_len}). "
+                    f"Dataset sample is too long ({seq_len} > {self.max_seq_len}). "
                     "Please set `split_across_pack=True` or increase `max_seq_len`."
                 )
 
-            # Create integer mask and position ids for current sample and extend
-            # current pack
-            current_sample = {
-                "tokens": tokens,
-                "labels": labels,
-                # Mask is simply a causal mask within this sample length
-                "mask": [torch.tril(torch.ones(seq_len, seq_len, dtype=torch.bool))],
-                "input_pos": list(range(seq_len)),
-            }
-            current_pack = {k: v + current_sample[k] for k, v in current_pack.items()}
+            # Update the current pack
+            current_pack["tokens"] += tokens
+            current_pack["labels"] += labels
+            current_pack["input_pos"] += list(range(seq_len))
+            current_pack["seq_lens"] += [seq_len]
 
             # If the current pack is long enough, add it to self.packs and retain
-            # any truncated samples for next pack, if splitting samples
-            if len(current_pack["tokens"]) > self.max_seq_len:
-                current_pack = self._add_pack(
-                    current_pack=current_pack,
-                    boundary=self.max_seq_len
-                    if self.split_across_pack
-                    else previous_sample_boundary,
-                )
+            # any truncated or bumped samples for next pack
+            if len(current_pack["tokens"]) >= self.max_seq_len:
+                current_pack = self._add_pack(current_pack)
 
             if rank == 0:
                 pbar.update()
-            previous_sample_boundary = len(current_pack["tokens"])
-            if self.max_packs is not None and len(self.packs) >= self.max_packs:
+
+            # Keep track of previous sample boundary
+            self.previous_sample_boundary = len(current_pack["tokens"])
+
+            # If max packs is set, stop packing when we reach that number
+            if self.max_packs is not None and len(self.packs) == self.max_packs:
                 break
 
-        # Add the last pack with remaining samples that did not fit in previous
-        if len(current_pack["tokens"]) > 0 and (
-            self.max_packs is None or len(self.packs) < self.max_packs
+        # Handle the last pack if there's leftover and we haven't filled up the max packs
+        if (
+            len(current_pack["tokens"]) > 0
+            and self.max_packs is not None
```

Comment:
if `self.max_packs is None` and we have a leftover pack, then this condition won't be true when we would want it to be, hence the `self.max_packs is None or`

Reply:
Ahh good catch - let me also make sure this is caught in a test.